gtsam/hybrid/DCSAM.h

```diff
@@ -0,0 +1,202 @@
+   * @param initialGuessDiscrete - Initial guess for discrete variables.
+   */
+  void update(const HybridNonlinearFactorGraph &graph,
+              const Values &initialGuessContinuous = Values(),
+              const DiscreteValues &initialGuessDiscrete = DiscreteValues());
+
+  /**
+   * Inline convenience function to allow "skipping" the initial guess for
+   * continuous variables while adding an initial guess for discrete variables.
+   */
+  inline void update(const HybridNonlinearFactorGraph &graph,
+                     const DiscreteValues &initialGuessDiscrete) {
+    update(graph, Values(), initialGuessDiscrete);
+  }
+
+  /**
+   * Simply used to call `update` without any new factors. Runs an iteration of
+   * optimization.
+   */
+  void update();
+
+  /**
+   * Add factors in `graph` to member discrete factor graph `dfg_`, then update
+   * any stored continuous variables using those in `values` by calling
+   * `updateDiscreteInfo(values)`.
+   *
+   * NOTE: could this be combined with `updateDiscreteInfo` or do these
+   * definitely need to be separate?
+   *
+   * @param dfg - a discrete factor graph containing the factors to add
+   * @param discreteValues - an assignment to the continuous variables
+   * (or subset thereof).
+   */
+  void updateDiscrete(const DiscreteFactorGraph &dfg = DiscreteFactorGraph(),
+                      const DiscreteValues &discreteVals = DiscreteValues());
+
+  /**
+   * For any factors in `dfg_`, update their stored local continuous information
+   * with the values from `values`.
+   *
+   * NOTE: could this be combined with `updateDiscrete` or do these
+   * definitely need to be separate?
+   *
+   * @param values - an assignment to the continuous variables (or subset
+   * thereof).
+   */
+  void updateDiscreteInfo(const DiscreteValues &discreteVals);
+
+  /**
+   * At the moment, this just calls `isam_.update()` internally
+   * This could, for example, be used if one wanted to hold the discrete
+   * variables constant and perform multiple iterations of updates to the
+   * continuous variables.
+   *
+   * NOTE: this function behaves quite differently from the similarly
+   * named `updateDiscrete` function, maybe we consider refactoring
+   * some of these function names.
+   */
+  void updateContinuous();
+
+  /**
+   * Given the latest discrete values (discreteVals),
+   * a set of new factors (newFactors),
+   * and an initial guess for any new keys (initialGuess),
+   * this function marks any affected keys as such, and calls
+   * `isam_.update` with the new factors and initial guess.
+   * See implementation for more detail.
+   */
+  void updateContinuousInfo(const DiscreteValues &discreteVals,
+                            const NonlinearFactorGraph &newFactors,
+                            const Values &initialGuess);
+
+  /**
+   * Solve for discrete variables given continuous variables. Internally, calls
+   * `dfg_.optimize()`
```

Comment:
This is not strictly accurate, I think, because the solve actually depends as well on the current state of the discrete variables.

For example, in the case of a max-mixture of hybrid factors (or an EM-factor of hybrid factors) the mode you're in (or mode probabilities) can depend on both the continuous assignment _and_ the discrete assignment.